src/datalab_app_plugin_insitu/utils.py

```diff
@@ -318,28 +337,134 @@ def _process_data(
-            nmr_data, df, num_experiments = process_pseudo2d_spectral_data(exp_folder)
+            nmr_data, df, num_experiments = process_pseudo2d_spectral_data(
+                exp_folder)
 
         else:
             raise ValueError(f"Unknown NMR dimension type: {nmr_dimension}")
 
         merged_df = (
-            process_echem_data(base_folder, echem_folder_name) if echem_folder_name else None
+            process_echem_data(
+                base_folder, echem_folder_name) if echem_folder_name else None
         )
 
-        return prepare_for_bokeh(nmr_data, df, merged_df, num_experiments)
+        result = prepare_for_bokeh(nmr_data, df, merged_df, num_experiments)
+
+        if result is None:
+            raise RuntimeError(
+                "prepare_for_bokeh returned None instead of expected data")
+
+        return result
 
     except Exception as e:
         raise RuntimeError(f"Error in common processing: {str(e)}")
+
+
+def get_cache_path(file_id: str) -> str:
+    """Get the path to the cache file for a given file_id."""
+    cache_dir = os.path.join(os.path.expanduser("~"), ".datalab_cache")
+    os.makedirs(cache_dir, exist_ok=True)
+
+    cache_path = os.path.join(cache_dir, f"{file_id}_insitu.h5")
+
+    return cache_path
+
+
+def save_to_cache(file_id: str, data: Dict) -> None:
+    """Save processed data to HDF5 cache file."""
+    cache_path = get_cache_path(file_id)
+
+    try:
+        os.makedirs(os.path.dirname(cache_path), exist_ok=True)
+
+        with h5py.File(cache_path, 'w') as f:
+            if "metadata" in data:
+                f.attrs["metadata"] = json.dumps(data["metadata"])
+
+            if "nmr_spectra" in data:
+                nmr_group = f.create_group("nmr_spectra")
+                nmr_group.create_dataset(
+                    "ppm", data=data["nmr_spectra"]["ppm"])
+
+                spectra_group = nmr_group.create_group("spectra")
+                for i, spectrum in enumerate(data["nmr_spectra"]["spectra"]):
+                    spectrum_group = spectra_group.create_group(str(i))
+                    spectrum_group.attrs["time"] = spectrum["time"]
+                    spectrum_group.attrs["experiment_number"] = spectrum.get(
+                        "experiment_number", i+1)
+                    spectrum_group.create_dataset(
+                        "intensity", data=spectrum["intensity"])
+
+            if "integrated_data" in data:
+                integrated_group = f.create_group("integrated_data")
+                for key, values in data["integrated_data"].items():
+                    integrated_group.create_dataset(key, data=values)
+
+            if "echem" in data and data["echem"]:
+                echem_group = f.create_group("echem")
+                for key, values in data["echem"].items():
+                    echem_group.create_dataset(key, data=values)
+
+        return True
+    except Exception as e:
+        raise RuntimeError(f"Error saving to cache: {str(e)}")
```

Comment:
Can we save the plot and the data in two separate files and load them selectively?